### PR TITLE
feat: EVA x HuggingFace

### PR DESCRIPTION
**Pull request by suryatejreddy**
This PR adds support for HuggingFace via [Pipelines](https://huggingface.co/docs/transformers/main_classes/pipelines).

### Background
The Pipeline API is an easy way to use a HuggingFace model for inference. In EVA, we require users to mention the task they want to perform for simplicity. A HuggingFace task is different from a model(pytorch). There are a large number of models on HuggingFace hub that can be used for a particular task. The user can specify the model or a default model will be used.

### Parser
We differentiate HF UDFs using the Type attribute of a UDF. It should be defined as "HuggingFace". The metadata attributes of the CREATE statement are used to specify/define the HF UDF. We expect "task" to be passed as metadata in the CREATE statement. Users can additionally specify "model" and other attributes specific to the model similarly. 

### Create UDF Executor
Currently, we do not support all tasks defined by HF. The supported tasks are defined in `eva/third_party/huggingface/create.py`. 

All HF UDFs have the same implementation file `udfs/abstract/hf_abstract_udf.py`. The inputs and outputs for the UDF are extracted using the following method:
- Identify the type of input for the UDF. This could be text, image, audio, or video. We store a 1-1 mapping between tasks dn input types. 
- Instantiate a pipeline using the attributes defined by the user (task, model, etc) and pass in a dummy input of correct type to return an output. We use this to identify the names and types of the outputs returned by the model. 
We generate IOCatalogEntry objects for these inputs and outputs to store along with the UDF. 

### Binder
The actual function for HF UDF is bound during the runtime of the function evaluation. We fetch the input to the UDF (generated in the above step). 

The data format for an HF UDF depends on the input type. We define different input types in `eva/third_party/huggingface/model.py`. Each implementation transforms the inputs into a format that can be passed to an HF model. 

### Final Syntax

```
CREATE UDF {udf_name}
TYPE HuggingFace
'task' 'text-classification'
```

```
CREATE UDF {udf_name}
TYPE HuggingFace
'task' 'object-detection'
'model' 'facebook / detr-resnet-50'
```

### Testing
Adds testcases

### CODE REVIEW

**Comment by gaurav274** [comment]
@suryatejreddy  Is it ready to review?

File: eva/binder/statement_binder.py [hunk 1]
```diff
@@ -243,40 +244,44 @@ def _bind_func_expr(self, node: FunctionExpression):
...
-        except Exception as e:
-            err_msg = (
-                f"{str(e)}. Please verify that the UDF class name in the"
-                "implementation file matches the UDF name."
-            )
-            logger.error(err_msg)
-            raise BinderError(err_msg)
+        if udf_obj.type == "HuggingFace":
+            node.function = bind_hf_func_from_udf(udf_obj)
+            node.projection_columns = ["*"]
```
Comment (gaurav274) [inline eva/binder/statement_binder.py:249]: What is this?

File: eva/utils/huggingface_utils.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    "object-detection": CustomObjectDetection,
+    "text-classification": CustomTextClassification,
+}
+
+
+def bind_hf_func_from_udf(udf_obj: UdfCatalogEntry):
+    """
+    Generate and return pipeline function from UdfCatalogEntry
+    """
+    default_args = {
```
Comment (gaurav274) [inline eva/utils/huggingface_utils.py:84]: Do you think we should default to this? Isn't it better to complain if they don't specify any task?

File: eva/utils/huggingface_utils.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+def image_data_preprocesser(images):
+    frames_list = images.values.tolist()
+    frames = np.array(frames_list)
+    frames = np.vstack(frames)
+    images = [Image.fromarray(row) for row in frames]
+    return images
+
+
+def data_postprocesser(model_output):
```
Comment (gaurav274) [inline eva/utils/huggingface_utils.py:36]: Same as above

File: eva/models/storage/batch.py [hunk 4]
```diff
@@ -247,6 +247,10 @@ def project(self, cols: None) -> Batch:
         We do a copy for now.
         """
         cols = cols or []
+        # The following is a workaround to support dynamic HuggingFace UDF's
+        # There is no concrete method of knowing the columns of the output beforehand
+        if cols == ["*"]:
```
Comment (gaurav274) [inline eva/models/storage/batch.py:252]: HuggingFace models cannot be composed in this case.

Comment (gaurav274) [reply eva/models/storage/batch.py:252]: @xzdandy Any ideas?

Comment (gaurav274) [reply eva/models/storage/batch.py:252]: Can't we get output based on the task?

File: docs/source/reference/hf.rst [hunk 5]
```diff
@@ -0,0 +1,35 @@
+HuggingFace Models in EVA
+======================
+
+This section provides an overview of how you can use out-of-thebox HuggingFace models in EVA.
```
Comment (gaurav274) [inline docs/source/reference/hf.rst:4]: "out-of-the box"

File: docs/source/reference/hf.rst [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+- Text Classification
+- Summarization
+- Text2Text Generation
+- Text Generation
+- Image Classification
+- Image Segmentation
+- Image-to-Text
+- Object Detection
+- Depth Estimation
+- Video Classification
```
Comment (gaurav274) [inline docs/source/reference/hf.rst:35]: We need to expand this. Maybe in the next PR.

File: eva/third_party/huggingface/binder.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from eva.catalog.catalog_type import NdArrayType
+from eva.catalog.models.udf_catalog import UdfCatalogEntry
+from eva.third_party.huggingface.model import ImageHFModel, TextHFModel
+
+
+def assign_hf_udf(udf_obj: UdfCatalogEntry):
+    # NOTE: Currently supports one input
```
Comment (gaurav274) [inline eva/third_party/huggingface/binder.py:21]: Please better explain this comment. Thanks!

Comment (gaurav274) [reply eva/third_party/huggingface/binder.py:21]: Add doc string

File: eva/third_party/huggingface/create.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+    return pipe(image=image_input, text=text_input)
+
+
+def infer_output_type(**pipeline_args):
+    """
+    Infer the output type of a pipeline based on the task type.
+    Note: The below code is flaky and depends on the version of transformers.
+    It also hard-codes the type assignment to certain tasks.
+    """
+    if "task" not in pipeline_args:
```
Comment (gaurav274) [inline eva/third_party/huggingface/create.py:104]: Convert all raised Exceptions to asserts.

File: eva/third_party/huggingface/create.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+    task = pipeline_args["task"]
+    if task in UNSUPPORTED_TASKS:
+        raise Exception(f"HuggingFace task:{task} not supported in EVA currently")
+
+    task_input = SUPPORTED_TASKS.get(task, {}).get("type", None)
+    if task_input is None:
+        raise Exception("Task Type Could Not Be Inferred")
+
+    pipe = pipeline(**pipeline_args)
+    if task_input == "text":
```
Comment (gaurav274) [inline eva/third_party/huggingface/create.py:115]: A lot of hard-coded strings. Will making them enums help? What is HF version specific here?

File: eva/udfs/abstract/hf_abstract_udf.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        return "GenericHuggingfaceModel"
+
+    def setup(
+        self, udf_obj: UdfCatalogEntry, device: int = -1, *args, **kwargs
+    ) -> None:
+        pipeline_args = {entry.key: entry.value for entry in udf_obj.metadata}
+        self.hf_udf_obj = pipeline(**pipeline_args, device=device)
+
+    @property
+    def input_format(self) -> str:
```
Comment (gaurav274) [inline eva/udfs/abstract/hf_abstract_udf.py:42]: What is this?

File: docs/source/reference/hf.rst (hunk 6 above)
Comment (suryatejreddy) [reply docs/source/reference/hf.rst:35]: Yeah the other models require more than one input so might require some more work. Will do that next.

Comment (gaurav274) [reply docs/source/reference/hf.rst:35]: I meant the documentation.

File: eva/third_party/huggingface/binder.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+def assign_hf_udf(udf_obj: UdfCatalogEntry):
+    """
+    Assigns the correct HF Model to the UDF. The model assigned depends on
+    the type of input the UDF expects. This is done so that we can
+    process the input correctly before passing it to the HF model.
+    """
+    inputs = udf_obj.args
+
+    # NOTE: Currently, we only support models that require a single input.
+    input_type = inputs[0].array_type
```
Comment (gaurav274) [inline eva/third_party/huggingface/binder.py:29]: Do we need an assert here for safety?

File: docs/source/reference/hf.rst (hunk 6 above)
Comment (suryatejreddy) [reply docs/source/reference/hf.rst:35]: What else do you think we can add? More examples?

File: eva/third_party/huggingface/create.py [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+from eva.catalog.catalog_type import ColumnType, NdArrayType
+from eva.catalog.models.udf_io_catalog import UdfIOCatalogEntry
+from eva.catalog.models.udf_metadata_catalog import UdfMetadataCatalogEntry
+from eva.third_party.huggingface.model import HFInputTypes
+
+"""
+We currently support the following tasks from HuggingFace.
+"""
+SUPPORTED_TASKS = {
+    "audio-classification": HFInputTypes.AUDIO,
```
Comment (gaurav274) [inline eva/third_party/huggingface/create.py:32]: Clean!

File: eva/third_party/huggingface/create.py [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+    "image-classification": HFInputTypes.IMAGE,
+    "image-segmentation": HFInputTypes.IMAGE,
+    "image-to-text": HFInputTypes.IMAGE,
+    "object-detection": HFInputTypes.IMAGE,
+    "depth-estimation": HFInputTypes.IMAGE,
+    "video-classification": HFInputTypes.VIDEO,
+}
+
+
+def run_pipe_through_text(pipe: Pipeline):
```
Comment (gaurav274) [inline eva/third_party/huggingface/create.py:48]: How about we have different functions to create the required data and decouple the `pipe(data)` call? I feel we will be able to clean it further.

File: eva/third_party/huggingface/create.py [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+    task = pipeline_args["task"]
+    assert task in SUPPORTED_TASKS, f"Task {task} not supported in EVA currently"
+
+    # Construct the pipeline
+    pipe = pipeline(**pipeline_args)
+
+    # Run the pipeline through a dummy input to get a sample output
+    model_input = SUPPORTED_TASKS[task]
+    if model_input == HFInputTypes.TEXT:
+        output = run_pipe_through_text(pipe)
```
Comment (gaurav274) [inline eva/third_party/huggingface/create.py:116]: Call `pipe(data)` here

File: eva/udfs/abstract/hf_abstract_udf.py [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+        result_df = pd.DataFrame(result_list)
+        return result_df
+
+    def forward(self, inputs, *args, **kwargs) -> None:
+        hf_input = self.input_formatter(inputs)
+        hf_output = self.hf_udf_obj(hf_input, *args, **kwargs)
+        eva_output = self.output_formatter(hf_output)
+        return eva_output
+
+    def to_device(self, device: str) -> GPUCompatible:
```
Comment (gaurav274) [inline eva/udfs/abstract/hf_abstract_udf.py:78]: Did you verify we are able to run it on GPUs?

File: eva/third_party/huggingface/create.py (hunk 13 above)
Comment (suryatejreddy) [reply eva/third_party/huggingface/create.py:48]: I initially did it that way, but the reason for not going ahead was for video. I would need a temp file for video and this was the cleanest way of dealing with videos.

File: docs/source/reference/hf.rst (hunk 6 above)
Comment (gaurav274) [reply docs/source/reference/hf.rst:35]: Yes

File: eva/udfs/abstract/hf_abstract_udf.py (hunk 15 above)
Comment (suryatejreddy) [reply eva/udfs/abstract/hf_abstract_udf.py:78]: Will cover in next PR with GPU redesign.

File: docs/source/reference/hf.rst [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+Creating UDF from HuggingFace
+-----
+EVA supports UDFS similar to `Pipelines <https://huggingface.co/docs/transformers/main_classes/pipelines>`_  in HuggingFace. 
+
+.. code-block:: sql
+
+    CREATE UDF IF NOT EXISTS HFObjectDetector
+    TYPE  HuggingFace
+    'task' 'object-detection'
+    'model' 'facebook / detr-resnet-50'
```
Comment (xzdandy) [inline docs/source/reference/hf.rst:16]: Is this another way to create UDF? Internally it will be translated to how EVA normally creates UDF.

Comment (suryatejreddy) [reply docs/source/reference/hf.rst:16]: Yeah that's the goal of the design, users can use HF UDF's without writing code and EVA treats it similar to any other UDF.

**Review by gaurav274 — APPROVED** [review]

**Pull request merged**